File: algorithms/74. Search a 2D Matrix/solution.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        m, n = len(matrix), len(matrix[0]) if matrix else 0
        l, r = 0, m * n - 1
        while l <= r:
            mid = (l + r) / 2
            num = matrix[mid / n][mid % n]
            if target < num:
                r = mid - 1
            elif target > num:
                l = mid + 1
            else:
                return True
        return False

File: algorithms/74. Search a 2D Matrix/solution.cc
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size(), n = m ? matrix[0].size() : 0;
        int l = 0, r = m * n - 1, mid, num;
        while ( l <= r ) {
            mid = (l + r) / 2;
            num = matrix[mid / n][mid % n];
            if ( target < num ) {
                r = mid - 1;
            } else if ( target > num ) {
                l = mid + 1;
            } else {
                return true;
            }
        }
        return false;
    }
};
